aexpect/ops.py

```diff
@@ -0,0 +1,479 @@
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError(f'Failed to create temporary directory: {output}')
+    return output
+
+
+def make_tempfile(session, template=None):
+    """
+    Create a temporary file in a safe way through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param template: file name template as used in the standard command
+    :type template: str or None
+    :returns: name of temporary file
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if mktemp returns non-zero exit status
+
+    Calls `mktemp`, refer to `man` for more info.
+    """
+    cmd = f'mktemp {template}' if template is not None else 'mktemp'
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError(f'Failed to create temporary file: {output}')
+    return output
+
+
+def cat(session, filename, quote_path=True, flags=""):
+    """
+    Get contents of a text file from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str filename: full path of file
+    :param bool quote_path: whether to quote (escape) or not (expand) the path
+    :param str flags: raw flags to pass to cat command
+    :returns: file contents
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if cat command fails
+
+    Should only be used for very short files without tabs or other fancy
+    contents. Otherwise better download file or use some other method.
+    """
+    cmd = f'cat {flags} {quote(filename)}' if quote_path else f'cat {flags} {filename}'
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError(f'Failed to cat {filename}: {output}')
+    return output
+
+
+def sed_replace(session, pattern, replacement, filename):
+    """
+    Replace a pattern in a file using sed, raising an error if nothing is found.
+
+    :param session: session of the remote host
+    :type session: :py:class:`RemoteSession`
+    :param str pattern: pattern to replace
+    :param str replacement: string to replace the pattern for
+    :param str filename: file with content to be replaced
+    :raises: :py:class:`AssertionError` if nothing was replaced
+
+    Using sed is convenient, but it does not error out when nothing is replaced
+    and this can lead to false-positives when used inside tests.
+    """
+    LOG.info("Replacing {pattern} with {replacement} in {filename}")
+    # for lines matching pattern, replace and print them (the p command won't work)
+    cmd = f"""sed -ri '/{pattern}/{{
+s//{replacement}/g
+w /dev/stdout
+}}' {filename}"""
+    output = session.cmd(cmd)
+    # make sure we have at least one line replaced
+    assert len(output.splitlines()) > 0, f"sed did not find any matches for pattern {pattern}"
```

Comment:
It's safer to raise exceptions as asserts are stripped when compiling with optimizations (although we do use it on a few places in aexpect already so take this as a mild recommendation)

Reply:
Also thank you for an interesting insight to verifying sed (although I never needed such feature it seems useful)

Reply:
Removing assertions for the purpose of optimization is something I typically strongly advise against, adding self-validation within the executing code is something very much needed to enhance the language's self-validation in these cases and and I doubt assertions like these are critical performance drag.

Regarding the self check of sed, this is another use we could have from such commands - use them to have additional sanity checks in place and post-processing that fits the aexpect specific usage.

Reply:
"Assertions are programmer aids, not error handlers for your business logic" - https://www.inspiredpython.com/article/assertions-and-exceptions-are-not-the-same

Reply:
I never intended to say they are the same but the current code does state otherwise and I moved this to similar RuntimeError handling logic with other functions.